### PR TITLE
[WIP] Adds reference tracking

### PR DESCRIPTION
**Pull request by MCHSL**
## About The Pull Request
Adds extools-powered reference tracking. Includes a couple procs that retrieve the back and forward references of a datum - Back references let you see what is referencing your object and potentially preventing it from garbage collecting, and forward ones show you what your object in turn references. Also made a cool GUI to inspect and follow these references.

image

The tracking adds some overhead to all variable sets and list operations. Init time is increased by ~15%. I haven't actually benched it so it might impact the actual game less.

## Why It's Good For The Game
no lagging caused by hard dels scanning the entire planet (once coders fix them)

### CODE REVIEW

**Comment by alexkar598** [comment]
This is pog

**Comment by kevinz000** [comment]
POG

**Comment by Qustinnus** [comment]
definitely pog 👍

File: code/modules/admin/view_variables/reference_tracking.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/proc/get_back_references(datum/D)
+	CRASH("/proc/get_back_references not hooked by extools, reference tracking will not function!")
+
+/proc/get_forward_references(datum/D)
+	CRASH("/proc/get_forward_references not hooked by extools, reference tracking will not function!")
+
+/proc/clear_references(datum/D)
+	return
+
+/datum/Del()
```
Comment (Fox-McCloud) [inline code/modules/admin/view_variables/reference_tracking.dm:15]: We absolutely cannot do this under any circumstance, ever.

If `Del` is even *defined* on an object, it dramatically increases the time/cost it takes to delete.

The fact this is defined at the `/datum` level means that now a ton of things are going to get incredibly more expensive to delete----and some things are just never going to GC (like mobs).

This is a great PR, but we can never, ever ever define `Del`.

Comment (AnturK) [reply code/modules/admin/view_variables/reference_tracking.dm:15]: Just needs an ifdef over the whole file. Same as as our current soft ref tracker.

**Comment by MCHSL** [comment]
Now extools hooks the function that handles deleting atoms (both del and soft GC hopefully) so no need for datum/Del. Init times are now increased by 10%.

**Comment by TerraGS** [comment]
reacted with the Pog emoji

**Comment by Cyberboss** [comment]
WIP?

**Comment by WarlockD** [comment]
Huh I learned something new called robin hood hashing.  The library is in extools.  Really neet idea.

**Pull request merged**